lib/BoundaryValueDiffEqFIRK/src/firk.jl

```diff
@@ -441,43 +441,65 @@ function __construct_nlproblem(
-            NoSparsityDetection()
-    cache_bc = __sparse_jacobian_cache(
-        Val(iip), jac_alg.bc_diffmode, sd_bc, loss_bcₚ, resid_bc, y)
+    cache_bc = if iip
+        DI.prepare_jacobian(loss_bc, resid_bc, bc_diffmode, y, Constant(cache.p))
+    else
+        DI.prepare_jacobian(loss_bc, bc_diffmode, y, Constant(cache.p))
+    end
 
-    sd_collocation = if jac_alg.nonbc_diffmode isa AutoSparse
+    nonbc_diffmode = if jac_alg.nonbc_diffmode isa AutoSparse
         if L < cache.M
             # For underdetermined problems we use sparse since we don't have banded qr
-            colored_matrix = __generate_sparse_jacobian_prototype(
-                cache, cache.problem_type, y, y, cache.M, N)
             J_full_band = nothing
-            __sparsity_detection_alg(ColoredMatrix(
-                sparse(colored_matrix.M), colored_matrix.row_colorvec,
-                colored_matrix.col_colorvec))
+            colored_result = __generate_sparse_jacobian_prototype(
+                cache, cache.problem_type, y, y, cache.M, N, jac_alg.nonbc_diffmode)
         else
             block_size = cache.M * (stage + 2)
             J_full_band = BandedMatrix(
                 Ones{eltype(y)}(L + cache.M * (stage + 1) * (N - 1),
                     cache.M * (stage + 1) * (N - 1) + cache.M),
                 (block_size, block_size))
-            __sparsity_detection_alg(__generate_sparse_jacobian_prototype(
-                cache, cache.problem_type, y, y, cache.M, N))
+            colored_result = __generate_sparse_jacobian_prototype(
+                cache, cache.problem_type, y, y, cache.M, N, jac_alg.nonbc_diffmode)
         end
+        AutoSparse(get_dense_ad(jac_alg.nonbc_diffmode);
+            sparsity_detector = ADTypes.KnownJacobianSparsityDetector(colored_result.A),
+            coloring_algorithm = ConstantColoringAlgorithm{ifelse(
+                ADTypes.mode(jac_alg.nonbc_diffmode) isa ADTypes.ReverseMode,
+                :row, :column)}(colored_result.A,
+                ifelse(ADTypes.mode(jac_alg.nonbc_diffmode) isa ADTypes.ReverseMode,
+                    row_colors, column_colors)(colored_result)))
     else
         J_full_band = nothing
-        NoSparsityDetection()
+        jac_alg.nonbc_diffmode
     end
 
-    cache_collocation = __sparse_jacobian_cache(
-        Val(iip), jac_alg.nonbc_diffmode, sd_collocation,
-        loss_collocationₚ, resid_collocation, y)
+    cache_collocation = if iip
+        DI.prepare_jacobian(
+            loss_collocation, resid_collocation, nonbc_diffmode, y, Constant(cache.p))
+    else
+        DI.prepare_jacobian(loss_collocation, nonbc_diffmode, y, Constant(cache.p))
+    end
 
-    J_bc = zero(init_jacobian(cache_bc))
-    J_c = zero(init_jacobian(cache_collocation))
+    J_bc = if iip
+        DI.jacobian(loss_bc, resid_bc, cache_bc, bc_diffmode, y, Constant(cache.p))
+    else
+        DI.jacobian(loss_bc, bc_diffmode, y, Constant(cache.p))
```

Comment:
```suggestion
        DI.jacobian(loss_bc, cache_bc, bc_diffmode, y, Constant(cache.p))
```
You can (and should) use the preparation result even when you compute the Jacobian out-of-place

Reply:
Fixed. Just out of curiosity, what's the difference between using or not using the preparation result when calling `DI.jacobian` on the OOP function? Using the preparation can make Jacobian computing much faster?